ssl/ssl_key_share.cc

```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
   HRSS_private_key hrss_private_key_;
 };
 
+// The hybrid key exchange implemented here is slightly different from the
+// key exchange used for CECPQ2. PQHybridKeyShare prepends 2 bytes to indicate
+// the length of each component public key immediately before the corresponding
+// key; CECPQ2 does not do this. PQHybridKeyShare is implemented according to:
+// https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design.
+class PQHybridKeyShare : public SSLKeyShare {
+ public:
+  PQHybridKeyShare(uint16_t group_id)
+      : group_id_(group_id) {
+    for (const PQGroup &pq_group : PQGroups()) {
+      if (group_id == pq_group.group_id) {
+        ec_nid_ = pq_group.ec_nid;
+        pq_nid_ = pq_group.pq_nid;
+        break;
+      }
+    }
+  }
+
+  uint16_t GroupID() const override { return group_id_; }
+
+  bool Offer(CBB *out) override {
+    assert(!pq_kem_ctx_);
+    assert(!ec_key_share_);
+
+    uint16_t ec_group_id;
+    uint16_t ec_public_key_length;
+    if (!ssl_nid_to_group_id(&ec_group_id, ec_nid_) ||
+        !get_ec_public_key_length(&ec_public_key_length, ec_nid_) ||
+        !(ec_key_share_ = SSLKeyShare::Create(ec_group_id)) ||
+        !CBB_add_u16(out, ec_public_key_length) || !ec_key_share_->Offer(out)) {
+      return false;
+    }
+
+    pq_kem_ctx_ = EVP_PQ_KEM_CTX_new();
+    if (!EVP_PQ_KEM_CTX_init_by_nid(pq_kem_ctx_, pq_nid_) ||
+        !EVP_PQ_KEM_generate_keypair(pq_kem_ctx_) ||
+        !CBB_add_u16(out, pq_kem_ctx_->kem->public_key_length) ||
+        !CBB_add_bytes(out, pq_kem_ctx_->public_key,
+                       pq_kem_ctx_->kem->public_key_length)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  bool Accept(CBB *out_public_key, Array<uint8_t> *out_secret,
+              uint8_t *out_alert, Span<const uint8_t> peer_key) override {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
```

Comment:
setting the alert here means the function will always return `SSL_AD_INTERNAL_ERROR` in `out_alert`, even on success. I don't think we should do this.

Reply:
Right, I was just trying to keep it consistent with what is already done in the other key shares
* https://github.com/awslabs/aws-lc/blob/f6614260186c1b437ae54406fcb9b7f0736b154b/ssl/ssl_key_share.cc#L79
* https://github.com/awslabs/aws-lc/blob/f6614260186c1b437ae54406fcb9b7f0736b154b/ssl/ssl_key_share.cc#L161